py3status/modules/speedtest.py

```diff
@@ -0,0 +1,296 @@
+
+STRING_NOT_INSTALLED = "not installed"
+MISSING_PLACEHOLDER = "{download} or {upload} placeholder required"
+
+
+class Py3status:
+    """
+    """
+
+    # available configuration parameters
+    button_refresh = 2
+    button_share = None
+    format = (
+        u"Speedtest [Up {upload} {upload_unit}] [Down {download} {download_unit}]"
+    )
+    server_id = None
+    si_units = False
+    sleep_timeout = 5
+    thresholds = {
+        "download": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+        "ping": [(200, "bad"), (150, "orange"), (100, "degraded"), (10, "good")],
+        "quality": [
+            ("ok", "good"),
+            ("bad", "degraded"),
+            ("faster", "good"),
+            ("slower", "degraded"),
+            ("faster", "good"),
+        ],
+        "upload": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+    }
+    timeout = 10
+    unit_bitrate = "MB/s"
+    unit_size = "MB"
+
+    def post_config_hook(self):
+        if not self.py3.check_commands("speedtest-cli"):
+            raise Exception(STRING_NOT_INSTALLED)
+
+        self.first_run = True
+
+        self.placeholders = self.py3.get_placeholders_list(self.format)
+
+        self.command = ["speedtest-cli --json --secure"]
+
+        # if download* or upload* missing, run complete test
+        if any(
+            key.startswith(x)
+            for x in ["download", "upload"]
+            for key in self.placeholders
+        ):
+            if not any(key.startswith("upload") for key in self.placeholders):
+                self.command += ["--no-upload"]
+
+            if not any(key.startswith("download") for key in self.placeholders):
+                self.command += ["--no-download"]
+
+        if self.server_id and int(self.server_id):
+            self.command += ["--server {}".format(self.server_id)]
+
+        if self.button_share and all(
+            x in self.placeholders for x in ["download", "upload"]
+        ):
+            self.command += ["--share"]
+
+        self.command += ["--timeout {}".format(self.timeout)]
+
+    def _is_running(self):
+        try:
+            self.py3.command_output(["pgrep", "speedtest-cli"])
+            return True
+        except self.py3.CommandError:
+            return False
+
+    def _get_speedtest_data(self):
+        command = ' '.join(self.command)
+        return loads(self.py3.command_output(command)) or None
+
+    def speedtest(self):
+        speedtest_data = {}
+        self.url = None
+        if self.first_run:
+            self.first_run = False
+            cached_until = self.py3.CACHE_FOREVER
+        else:
+            current_data = {}
+            previous_data = {}
+            cached_until = self.py3.time_in(self.sleep_timeout)
+
+            if not self._is_running():
+                previous_data = self.py3.storage_get("speedtest_data")
+                current_data = self._get_speedtest_data()
+                cached_until = self.py3.CACHE_FOREVER
```

Comment:
We don't need three `cached_until`. Move this to top.

Reply:
seen